src/OpenTelemetry.Contrib.Instrumentation.GrpcCore/RpcScope.cs

```diff
@@ -0,0 +1,231 @@
+        /// </summary>
+        private readonly bool recordMessageEvents;
+
+        /// <summary>
+        /// The RPC activity.
+        /// </summary>
+        private Activity activity;
+
+        /// <summary>
+        /// The complete flag.
+        /// </summary>
+        private long complete = 0;
+
+        /// <summary>
+        /// The request message counter.
+        /// </summary>
+        private int requestMessageCounter;
+
+        /// <summary>
+        /// The response counter.
+        /// </summary>
+        private int responseMessageCounter;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="RpcScope{TRequest, TResponse}" /> class.
+        /// </summary>
+        /// <param name="fullServiceName">Full name of the service.</param>
+        /// <param name="recordMessageEvents">if set to <c>true</c> [record message events].</param>
+        protected RpcScope(string fullServiceName, bool recordMessageEvents)
+        {
+            this.FullServiceName = fullServiceName?.TrimStart('/') ?? "unknownservice/unknownmethod";
+            this.recordMessageEvents = recordMessageEvents;
+        }
+
+        /// <summary>
+        /// Gets the full name of the service.
+        /// </summary>
+        protected string FullServiceName { get; }
+
+        /// <summary>
+        /// Records a request message.
+        /// </summary>
+        /// <param name="request">The request.</param>
+        public void RecordRequest(TRequest request)
+        {
+            this.requestMessageCounter++;
+
+            if (this.activity == null || !this.activity.IsAllDataRequested || !this.recordMessageEvents)
+            {
+                return;
+            }
+
+            this.AddMessageEvent(typeof(TRequest).Name, request as IMessage, request: true);
+        }
+
+        /// <summary>
+        /// Records a response message.
+        /// </summary>
+        /// <param name="response">The response.</param>
+        public void RecordResponse(TResponse response)
+        {
+            this.responseMessageCounter++;
+
+            if (this.activity == null || !this.activity.IsAllDataRequested || !this.recordMessageEvents)
+            {
+                return;
+            }
+
+            this.AddMessageEvent(typeof(TResponse).Name, response as IMessage, request: false);
+        }
+
+        /// <summary>
+        /// Completes the RPC.
+        /// </summary>
+        public void Complete()
+        {
+            if (this.activity == null)
+            {
+                return;
+            }
+
+            // The overall Span status should remain unset however the grpc status code attribute is required
+            this.StopActivity((int)Grpc.Core.StatusCode.OK);
+        }
+
+        /// <summary>
+        /// Records a failed RPC.
+        /// </summary>
+        /// <param name="exception">The exception.</param>
+        public void CompleteWithException(Exception exception)

```

Comment:
No call to `activity.RecordException`?

Reply:
if that is part of the OpenTelemetry Activity contract I can do that. Is it?

Reply:
If you have a dependency on API then you'll have this extension: https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry.Api/Trace/ActivityExtensions.cs#L77